### PR TITLE
Optimize performance by fuse adding high precision tensor to fp8 tensor

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Optimize performance by fuse add high precision tensor to fp8 tensor

**Major Revision**
- Add  an extension  msamp_arithmetic
- Add fuse kernel for adding high precision tensor to FP8 in extension
- Move common header files to msamp/common/include
- Add UT
- Apply it megatron-FP8DistributedDataParallel

### CODE REVIEW

**Review by wkcn — APPROVED** [review]
LGTM. Thanks!

**Review by wkcn — COMMENTED** [review]
Please declare the source and license of msamp/common/include/concurrency.h and msamp/common/include/poll.h

**Comment by tocean** [comment]
> Please declare the source and license of msamp/operators/arithmetic/vectorized_pointwise.h, msamp/common/include/concurrency.h and msamp/common/include/poll.h
> 
> An example: Adapted from xxx, xxx license.

Good suggestion. Fixed.

File: msamp/common/include/common.h [hunk 1]
```diff
@@ -68,4 +106,24 @@ template <> __host__ __device__ bf16 cast_dtype(const float value) { return __fl
...
+struct is_fp8<fp8e5m2> : std::true_type {};
+
+template <typename T>
+struct is_fp16 : std::false_type {};
+
+template <>
+struct is_fp16<fp16> : std::true_type {};
+
+template <>
+struct is_fp16<bf16> : std::true_type {};
```
Comment (wkcn) [inline msamp/common/include/common.h:127]: It is ambiguous that bf16 is fp16. I prefer to rename it to `is_fp16_or_bf16`

File: tests/operators/test_arithmetic.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        assert torch.all(torch.eq(scaling_tensor1.value, scaling_tensor2.value))
+        assert torch.all(torch.eq(scaling_tensor1.meta.scale, scaling_tensor2.meta.scale))
+        assert torch.all(torch.eq(scaling_tensor1.meta.scale_inv, scaling_tensor2.meta.scale_inv))
+        assert torch.all(torch.eq(scaling_tensor1.meta.amax, scaling_tensor2.meta.amax))
+
+    @decorator.cuda_test
+    def test_add_to_fp8(self):
+        """Test the function Arithmetic.add_to_fp8()."""
+        torch.manual_seed(100)
+        sizes = []
```
Comment (wkcn) [inline tests/operators/test_arithmetic.py:28]: The code L28-31 can be removed.

File: tests/operators/test_arithmetic.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+from tests.helper import decorator
+from msamp.common.dtype import Dtypes
+from msamp.operators.arithmetic import Arithmetic
+
+
+class ArithmeticTestCase(unittest.TestCase):
+    """A class for Arithmetic test cases."""
+    def _check_scaling_tensor(self, scaling_tensor1, scaling_tensor2):
+        assert torch.all(torch.eq(scaling_tensor1.value, scaling_tensor2.value))
```
Comment (wkcn) [inline tests/operators/test_arithmetic.py:19]: It is better to use `self.assertTrue`

**Comment by tocean** [comment]
Verified correctness using GPT3-345m:

baseline:
------------------------------------------------------------------------------------------------
 validation loss at iteration 1000 | lm loss value: 6.059220E+00 | lm loss PPL: 4.280416E+02 | 
------------------------------------------------------------------------------------------------

exp:
------------------------------------------------------------------------------------------------
 validation loss at iteration 1000 | lm loss value: 6.059220E+00 | lm loss PPL: 4.280416E+02 | 
------------------------------------------------------------------------------------------------

File: tests/operators/test_arithmetic.py (hunk 2 above)
Comment (tocean) [reply tests/operators/test_arithmetic.py]: Fixed.

File: tests/operators/test_arithmetic.py (hunk 3 above)
Comment (tocean) [reply tests/operators/test_arithmetic.py]: Good point. Fixed.

File: msamp/common/include/common.h (hunk 1 above)
Comment (tocean) [reply msamp/common/include/common.h]: Good point. Maybe `is_half` is better.

**Review by penghouwen — APPROVED** [review]

**Pull request merged**